### ISSUE TITLE
Integrating finite element method solvers within differentiable models by bridging JAX and FEniCS

### ISSUE THREAD

**Issue by matt-graham** [CCMI-CDT/programming-projects#4]
### Your name, department, and University

Matt Graham, Centre for Advanced Research Computing, UCL

### Name(s) and department(s) of anyone else relevant to this project

Matthew Scroggs, Centre for Advanced Research Computing, UCL

### Please write a brief description of the application area of project

Partial differential equations (PDEs) can be used to describe a wide range of different processes which vary in space and/or time. The finite element method (FEM) is one class of approaches for numerically solving PDEs. Given a PDE and appropriate boundary conditions, we can use the FEM to solve for the values of the unknown variables across space and time.

While in some cases the solution to a PDE is the end goal, in other situations it is an intermediate step in a larger computation. For instance, in PDE constrained inverse problems, the aim is to estimate the parameters of a PDE given noisy and partial observations of its solution. This requires solving a PDE with candidate parameter values and calculating the misfit between the observed values and generated solution. PDE solvers may also be included as part of larger 'hybrid' models combining physics-based and data-driven machine learning components. In both these settings being able to compute the derivatives of functions of the solution with respect to parameters of the model is required to allow using efficient gradient-based methods for fitting models to data.

### Please describe the project.

The aim of this project will be to allow including FEM solvers from [the open-source library FEniCS](https://fenicsproject.org/) within larger differentiable pipelines in the [high-performance array computing framework JAX](https://docs.jax.dev/en/latest/). This will build on previous work done in [a Python package `jax-fenics`](https://github.com/IvanYashchuk/jax-fenics) which implemented a prototype of precisely this functionality but depends on now unsupported versions of JAX and FEniCS and is no longer actively developed.

JAX is widely used in both machine learning and scientific computing contexts and there are a wide range of libraries in [the JAX ecosystem](https://docs.jax.dev/en/latest/#ecosystem), for example for constructing neural network models, performing optimization and probabilistic programming. Providing an easy to use bridge between JAX and FEniCS will allow simple integration of PDE models with these downstream libraries - for example for inferring the posterior in a PDE constrained inverse problem using approximate Bayesian inference algorithms in [PyMC](https://www.pymc.io/) or [Numpyro](https://num.pyro.ai/en/latest/index.html) or using a PDE model as part of a physics informed neural network in [Flax ](https://flax.readthedocs.io/) or [Equinox](https://docs.kidger.site/equinox/).

### What will be the outputs of this project?

The key output of the project will be an open-source Python package providing support for wraping FEniCSx solve and assemble operations as differentiable JAX primitives which works with recently released JAX and FEniCS(x) versions. This may involve upstream contributions to the [in progress DOLFINx adjoint package `dxa`](https://github.com/jorgensd/dxa) which is aiming to provide support for using the adjoint method, closely related to reverse-mode algorithmic / automatic differentiation, for DOLFINx solver operations. As additional goals, the package may also provide support for additional JAX function transformations such as automatic batching and just-in-time compilation, but the initial focus will be on support for automatic differentiation.

### Which programming language(s) will this project use?

Python, C++

### Links to any relevant existing software repositories, etc.

jax-fenics: https://github.com/IvanYashchuk/jax-fenics
DOLFINx: https://github.com/FEniCS/dolfinx
JAX: https://github.com/jax-ml/jax/
dxa: https://github.com/jorgensd/dxa

### Make project public

- [x] I understand that this project proposal will be public

**Comment by skyepurchase** [comment]
I would keen to work on this. I have some other JAX proposals that I will add soon

**Comment by bryceshirley** [comment]
I would also be keen to work on this.

**Comment by bryceshirley** [comment]
What are expectations in terms of time to work on this project?

**Comment by matt-graham** [comment]
Hi @bryceshirley, that's a good question. After some discussion with @mscroggs I think I wouldn't be planning to start this project at the upcoming software week beginning 3rd November, as I am going to be away at workshop for part of that week, so most likely this would start at the software week in the following term, or could potentially start before that in ~December, if there is interest. 

In terms of actual amount of time required to work on the project - this would be very much up to how much anyone who get's involved wants to commit. I think with a small team of 2-3 people we could get an initial minimal package together (basic API for differentiating through linear and non-linear solves with associated tests and API documentation) with a few days input from each person. If people are interested though this could be extended to a longer term project as there lots of ways I think we could expand this and make it more useable, not least adding a range of interesting tutorial examples.
